core/trino-main/src/main/java/io/trino/sql/planner/PlanFragmenter.java

```diff
@@ -356,21 +358,29 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
                 context.get().setDistribution(partitioningScheme.getPartitioning().getHandle(), metadata, session);
             }
 
-            ImmutableList.Builder<SubPlan> builder = ImmutableList.builder();
+            ImmutableList.Builder<FragmentProperties> childrenProperties = ImmutableList.builder();
+            ImmutableList.Builder<SubPlan> childrenBuilder = ImmutableList.builder();
             for (int sourceIndex = 0; sourceIndex < exchange.getSources().size(); sourceIndex++) {
                 FragmentProperties childProperties = new FragmentProperties(partitioningScheme.translateOutputLayout(exchange.getInputs().get(sourceIndex)));
-                builder.add(buildSubPlan(exchange.getSources().get(sourceIndex), childProperties, context));
+                childrenProperties.add(childProperties);
+                childrenBuilder.add(buildSubPlan(exchange.getSources().get(sourceIndex), childProperties, context));
             }
 
-            List<SubPlan> children = builder.build();
+            List<SubPlan> children = childrenBuilder.build();
             context.get().addChildren(children);
 
             List<PlanFragmentId> childrenIds = children.stream()
                     .map(SubPlan::getFragment)
                     .map(PlanFragment::getId)
                     .collect(toImmutableList());
 
-            return new RemoteSourceNode(exchange.getId(), childrenIds, exchange.getOutputSymbols(), exchange.getOrderingScheme(), exchange.getType());
+            return new RemoteSourceNode(
+                    exchange.getId(),
+                    childrenIds,
+                    exchange.getOutputSymbols(),
+                    exchange.getOrderingScheme(),
+                    exchange.getType(),
+                    isWorkerCoordinatorBoundary(context.get(), childrenProperties.build()) ? getRetryPolicy(session) : RetryPolicy.NONE);
```

Comment:
What happens if there's a failure in one of the upstream tasks? (I haven't read the other commits, so maybe the answer is clear later on)

Reply:
The failure will be safely ignored by the `DeduplicationExchangeClientBuffer`. Coordinator will submit a new attempt. Based on the attempt id passed to the `DeduplicationExchangeClientBuffer` the output of the old task will be safely discarded.